apps/next-evans/lib/db/friends.ts

```diff
@@ -27,40 +25,108 @@ export async function getAllFriends(
-  lang: Lang = LANG,
-): Promise<Record<string, DocumentWithMeta>> {
-  const friends = await getAllFriends(lang);
-  const documents: Record<string, DocumentWithMeta> = {};
-  Object.values(friends).forEach((friend) => {
-    friend.documents.forEach((doc) => {
-      documents[`${friend.slug}/${doc.slug}`] = {
-        ...doc,
-        publishedRegion: getPublicationRegion(friend.residences),
-        publishedDate: getPublicationDate(friend.residences, friend.born, friend.died),
-        authorGender: friend.gender,
-        authorName: friend.name,
-        authorSlug: friend.slug,
+async function getFriendsFromDB(lang: Lang): Promise<Record<string, Friend>> {
+  const publishedFriends = await queryFriends(lang);
+  const friendProps = publishedFriends
+    .map((friend) => {
+      return {
+        ...friend,
+        quotes: friend.friend_quotes.map((q) => ({ quote: q.text, cite: q.source })),
+        created_at: null,
+        dateAdded: friend.created_at.toISOString(),
+        residences: friend.friend_residences.map((r) => ({
+          city: r.city,
+          region: r.region,
+          durations: r.friend_residence_durations,
+        })),
+        documents: friend.documents.map((doc) => {
+          const firstEdition = doc.editions[0];
+          invariant(firstEdition !== undefined);
+          invariant(firstEdition.edition_impressions !== null);
+          const newestEdition = doc.editions.find(
+            (e) => e.type === mostModernEdition(doc.editions.map((ed) => ed.type)),
+          );
+          invariant(newestEdition !== undefined);
+
+          return {
+            ...doc,
+            altLanguageId: doc.alt_language_id,
+            created_at: null,
+            isbn: firstEdition.isbns[0]?.code ?? ``,
+            dateAdded: doc.created_at.toISOString(),
+            featuredDescription: doc.featured_description,
+            mostModernEdition: toEdition(newestEdition),
+            editions: doc.editions.map(toEdition),
+            shortDescription: doc.partial_description,
+            hasAudio: doc.editions.some((e) => e.edition_audios?.id),
+            tags: doc.document_tags.map((t) => t.type),
+            numDownloads: doc.editions.reduce((acc, ed) => acc + ed.downloads.length, 0), // TODO ~ maybe figure out if prisma can just count the downloads so don't have to get the ids
+            customCSS: null,
+            customHTML: null,
+          };
+        }),
       };
-    });
-  });
-  return documents;
+    })
+    .filter((friend) => friend.documents.length > 0);
+
+  return friendProps.reduce<Record<string, Friend>>((acc, friend) => {
+    acc[friend.slug] = friend;
+    return acc;
+  }, {});
 }
 
-export async function getNumDocuments(lang: Lang): Promise<number> {
-  return Object.values(await getAllDocuments(lang)).length;
+function toEdition(
+  apiEdition: Awaited<
+    ReturnType<typeof queryFriends>
+  >[number]['documents'][number]['editions'][number],
```

Comment:
i think it would be preferable to name this monstrous type outside the function definition. and i would propose maybe calling it `DbEdition` or  `QueryEdition` rather than "api-" because it's not actually coming from our api, it's just coming from the database. so the variable name `apiEdition` maybe should be changed too.